java/src/main/java/actorworkflow/CustomerLoyaltyWorkflowImpl.java

```diff
@@ -0,0 +1,147 @@
+
+import java.time.Duration;
+import java.util.ArrayList;
+
+public class CustomerLoyaltyWorkflowImpl implements CustomerLoyaltyWorkflow {
+
+    private static final Logger logger = Workflow.getLogger(CustomerLoyaltyWorkflowImpl.class);
+
+    private final CustomerLoyaltyActivities activities =
+            Workflow.newActivityStub(
+                    CustomerLoyaltyActivities.class,
+                    ActivityOptions.newBuilder()
+                            .setStartToCloseTimeout(Duration.ofSeconds(2))
+                            .build());
+
+    boolean accountActive = true;
+
+    Customer customer;
+
+    @Override
+    public String customerLoyalty(Customer customer) {
+        System.out.println("Customer passed in is: " + customer);
+        this.customer = customer;
+        WorkflowInfo info = Workflow.getInfo();
+
+        logger.info("Started workflow. Customer: {}.", this.customer);
+
+        if (info.getContinuedExecutionRunId().isEmpty()) {
+            String tier = this.customer.status().name();
+            activities.sendEmail(EmailStrings.EMAIL_WELCOME.formatted(tier));
+        }
+
+        // block on everything
+        while (true) {
+            Workflow.await(() -> !accountActive || info.getHistoryLength() > Shared.HISTORY_THRESHOLD);
+            if (accountActive) {
+                logger.info("Account still active, history size crossed limit; continuing-as-new.");
+                Workflow.continueAsNew(this.customer);
+            } else {
+                logger.info("Account canceled. Closing workflow.");
+                return "Done";
+            }
+        }
+    }
+
+    @Override
+    public void addLoyaltyPoints(int pointsToAdd) {
+        customer = customer.withPoints(customer.loyaltyPoints() + pointsToAdd);
+        logger.info("Added {} points to customer. Total loyalty points now {}.",
+                pointsToAdd, customer.loyaltyPoints());
+
+        StatusTier tierToPromoteTo = StatusTier.getMaxTier(customer.loyaltyPoints());
+
+        if (customer.status().minimumPoints() < tierToPromoteTo.minimumPoints()) {
+            logger.info("Promoting customer!");
+            customer = customer.withStatus(tierToPromoteTo);
+            activities.sendEmail(EmailStrings.EMAIL_PROMOTED.formatted(tierToPromoteTo.name()));
+        }
+    }
+
+    @Override
+    public void inviteGuest(Customer guest) {
+        logger.info("Checking to see if customer can invite guests.");
+        if (Customer.canAddGuest(customer)) {
+            logger.info("Customer is allowed to invite guests; attempting to start workflow for guest ID {}.",
+                    guest.customerId());
+            customer.guests().add(guest);
+
+            StatusTier guestMinStatus = StatusTier.previous(customer.status());
+            guest = guest.withStatus(guestMinStatus);
+
+            String guestWorkflowId = Shared.WORKFLOW_ID_FORMAT.formatted(guest.customerId());
+            ChildWorkflowOptions options =
```

Comment:
I think for this whole use case of starting child workflow async with abandoned mode and checking if its started and if yes then send signals would be easier to just invoke activity that then does SignalWithStart this workflow execution

Reply:
I didn't do that initially because there's it removes the parent/child relationship. In retrospect, since I'm abandoning anyways, it's probably not important to maintain that link. I'll look at doing activity -> signalwithstart instead.